In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
import pandas as pd

raw_data = pd.read_csv('/content/drive/My Drive/PCP/Star99999_raw.csv')
raw_data.head(5)

Unnamed: 0   Vmag      Plx   e_Plx     B-V SpType
0           0   9.10     3.54    1.39   0.482     F5
1           1   9.27    21.90    3.10   0.999    K3V
2           2   6.61     2.81    0.63  -0.019     B9
3           3   8.06     7.75    0.97   0.370    F0V
4           4   8.55     2.87    1.11   0.902  G8III

In [ ]:
type(raw_data)

pandas.core.frame.DataFrame

# Preprocesando el StarDataSet


# Limpieza de datos

Aquí puedes encontrar muchos más ejemplos:
* [Data Cleaning course on Kaggle Learn](https://www.kaggle.com/alexisbcook/handling-missing-values)

### Echemos un primer vistazo a los datos
Lo primero que tendremos que hacer es
* Importar bibliotecas
* Consulta los archivos que tenemos
* Cargar el conjunto de datos sin procesar

In [ ]:
import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

In [ ]:
# Use pd.read_csv to read file
path = '/content/drive/My Drive/PCP/Star99999_raw.csv'
raw_data = pd.read_csv(path)
raw_data.head(5)

Unnamed: 0   Vmag      Plx   e_Plx     B-V SpType
0           0   9.10     3.54    1.39   0.482     F5
1           1   9.27    21.90    3.10   0.999    K3V
2           2   6.61     2.81    0.63  -0.019     B9
3           3   8.06     7.75    0.97   0.370    F0V
4           4   8.55     2.87    1.11   0.902  G8III

Podemos ver que el conjunto de datos incluye una columna de índice duplicada, por lo que debemos eliminarlo más tarde.

Primero leamos algunas estadísticas del conjunto de datos.

In [ ]:
raw_data.columns

Index(['Unnamed: 0', 'Vmag', 'Plx', 'e_Plx', 'B-V', 'SpType'], dtype='object')

* `Vmag` - Magnitud visual aparente de la estrella
* `Plx` - Distancia entre la estrella y la Tierra
* ʻE_Plx` - Error estándar de `Plx` (¡Elimine la fila si encuentra que e_Plx es demasiado alto!)
* `B-V`: índice de color B-V. (Una estrella caliente tiene un índice de color B-V cercano a 0 o negativo, mientras que una estrella fría tiene un índice de color B-V cercano a 2.0. Otras estrellas están en algún punto intermedio).
* `SpType` - [Clasificación estelar.](https://en.wikipedia.org/wiki/Stellar_classification) 


In [ ]:
# read some statistics of the dataset
raw_data.describe()

Unnamed: 0
count  99999.000000
mean   49999.000000
std    28867.369122
min        0.000000
25%    24999.500000
50%    49999.000000
75%    74998.500000
max    99998.000000

¿Por qué la función de descripción no devuelve un resumen de todas las columnas?

Probablemente se deba a que este marco de datos tiene tipos de columna mixtos. El comportamiento predeterminado de la función de descripción de pandas es solo proporcionar un resumen de las columnas numéricas.

In [ ]:
# Check the DataType of our dataset
raw_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99999 entries, 0 to 99998
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Unnamed: 0  99999 non-null  int64 
 1   Vmag        99999 non-null  object
 2   Plx         99999 non-null  object
 3   e_Plx       99999 non-null  object
 4   B-V         99999 non-null  object
 5   SpType      97377 non-null  object
dtypes: int64(1), object(5)
memory usage: 4.6+ MB


^ Como podemos ver, tanto **`Vmag`,` Plx`, ʻe_Plx`, `B-V`** están marcados como objeto, pero se suponía que tenían un valor flotante.

Entonces deberíamos convertir nuestras columnas a numéricas.

Si ejecutamos este código:
```python
# Convert Columns data type to float values
raw_data["Vmag"] = pd.to_numeric(raw_data["Vmag"], downcast="float")
raw_data["Plx"] = pd.to_numeric(raw_data["Plx"], downcast="float")
raw_data["e_Plx"] = pd.to_numeric(raw_data["e_Plx"], downcast="float")
raw_data["B-V"] = pd.to_numeric(raw_data["B-V"], downcast="float")
```
Error would Occur : `ValueError: Unable to parse string "     " at position 25189`

Desde el error anterior, podemos observar que algunas celdas se llenaron con espacios en blanco, por lo que no se pueden analizar.

Podemos agregar un parámetro ʻerrores = 'coaccionar'` para forzar a la función a convertir valores no numéricos incorrectos a NaN.

https://stackoverflow.com/questions/40790031/pandas-to-numeric-find-out-which-string-it-was-unable-to-parse

In [ ]:
raw_data["Vmag"] = pd.to_numeric(raw_data["Vmag"], downcast="float")
raw_data["Plx"] = pd.to_numeric(raw_data["Plx"], downcast="float")
raw_data["e_Plx"] = pd.to_numeric(raw_data["e_Plx"], downcast="float")
raw_data["B-V"] = pd.to_numeric(raw_data["B-V"], downcast="float")

ValueError: ignored

In [ ]:
# Convert Columns data type to float values
raw_data["Vmag"] = pd.to_numeric(raw_data["Vmag"], downcast="float", errors='coerce')
raw_data["Plx"] = pd.to_numeric(raw_data["Plx"], downcast="float", errors='coerce')
raw_data["e_Plx"] = pd.to_numeric(raw_data["e_Plx"], downcast="float", errors='coerce')
raw_data["B-V"] = pd.to_numeric(raw_data["B-V"], downcast="float", errors='coerce')

Ahora revisemos la información nuevamente.

In [ ]:
# Check the DataType of our dataset
raw_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99999 entries, 0 to 99998
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Unnamed: 0  99999 non-null  int64  
 1   Vmag        99998 non-null  float32
 2   Plx         99768 non-null  float32
 3   e_Plx       99768 non-null  float32
 4   B-V         98871 non-null  float32
 5   SpType      97377 non-null  object 
dtypes: float32(4), int64(1), object(1)
memory usage: 3.1+ MB


In [ ]:
raw_data.describe()

Unnamed: 0          Vmag           Plx         e_Plx           B-V
count  99999.000000  99998.000000  99768.000000  99768.000000  98871.000000
mean   49999.000000      8.369750      7.212467      1.365385      0.704726
std    28867.369122      1.313896     11.348988      1.816838      0.489679
min        0.000000     -1.440000    -54.950001      0.380000     -0.400000
25%    24999.500000      7.640000      2.510000      0.880000      0.348000
50%    49999.000000      8.440000      4.630000      1.100000      0.612000
75%    74998.500000      9.140000      8.410000      1.390000      1.075000
max    99998.000000     14.080000    772.330017    114.459999      5.460000

In [ ]:
# Actually , if you want to show all the columns you can add parameter `include='all'`.
raw_data.describe(include='all')

Unnamed: 0          Vmag  ...           B-V  SpType
count   99999.000000  99998.000000  ...  98871.000000   97377
unique           NaN           NaN  ...           NaN    3756
top              NaN           NaN  ...           NaN      K0
freq             NaN           NaN  ...           NaN    7355
mean    49999.000000      8.369750  ...      0.704726     NaN
std     28867.369122      1.313896  ...      0.489679     NaN
min         0.000000     -1.440000  ...     -0.400000     NaN
25%     24999.500000      7.640000  ...      0.348000     NaN
50%     49999.000000      8.440000  ...      0.612000     NaN
75%     74998.500000      9.140000  ...      1.075000     NaN
max     99998.000000     14.080000  ...      5.460000     NaN

[11 rows x 6 columns]

## Revisando datos nulos
Es muy común que un conjunto de datos tenga algunos valores nulos.

In [ ]:
# obtener el numero de datos vacíos por columna
missing_values_count = raw_data.isnull().sum()

missing_values_count

Unnamed: 0       0
Vmag             1
Plx            231
e_Plx          231
B-V           1128
SpType        2622
dtype: int64

In [ ]:
raw_data.shape

(99999, 6)

In [ ]:
# ¿cuantos valores vacíos tenemos?
total_cells = np.product(raw_data.shape)
total_missing = missing_values_count.sum()

# porcentaje de data perdida
percent_missing = (total_missing/total_cells)
print("Percentage Missing:", "{:.2%}".format(percent_missing))

Percentage Missing: 0.70%


A partir del porcentaje de datos que faltan, dado que es tan pequeño (solo 0,7%), podemos eliminar las filas.

Veamos qué pasará si eliminamos todas las filas que contienen un valor faltante.

## Eliminando data faltante

In [ ]:
# eliminar todas las filas que contienen un valor faltante
# mejor almacenarlo en una nueva variable para evitar confusiones
raw_data_na_dropped = raw_data.dropna() 

raw_data_na_dropped

Unnamed: 0  Vmag    Plx  e_Plx    B-V    SpType
0               0  9.10   3.54   1.39  0.482        F5
1               1  9.27  21.90   3.10  0.999       K3V
2               2  6.61   2.81   0.63 -0.019        B9
3               3  8.06   7.75   0.97  0.370       F0V
4               4  8.55   2.87   1.11  0.902     G8III
...           ...   ...    ...    ...    ...       ...
99993       99993  7.92   3.93   1.24  0.133     Am...
99994       99994  8.72   3.07   0.87  0.097        B3
99996       99996  8.08   1.07   0.68  1.094        G5
99997       99997  6.98   2.97   0.76 -0.143     B1.5V
99998       99998  8.51  -1.18   1.34  1.568  K5/M0III

[96742 rows x 6 columns]

In [ ]:
# ¿Cuántas filas eliminaremos?
dropped_rows_count = raw_data.shape[0]-raw_data_na_dropped.shape[0]
print("filas eliminadas del dataset original %d \n" % dropped_rows_count)

# Porcentaje de filas eliminadas
percent_dropped = dropped_rows_count/raw_data.shape[0]
print("Porcentaje perdido:", "{:.2%}".format(percent_dropped))

filas eliminadas del dataset original 3257 

Porcentaje perdido: 3.26%


Por último, vuelva a leer las estadísticas y la información.

In [ ]:
raw_data_na_dropped.describe()

Unnamed: 0          Vmag           Plx         e_Plx           B-V
count  96742.000000  96742.000000  96742.000000  96742.000000  96742.000000
mean   50008.472721      8.291918      7.105242      1.264163      0.700663
std    28877.277238      1.247995     11.145643      1.156202      0.490053
min        0.000000     -1.440000    -35.099998      0.380000     -0.400000
25%    25001.250000      7.610000      2.520000      0.870000      0.342000
50%    49971.500000      8.400000      4.600000      1.080000      0.605000
75%    75034.750000      9.070000      8.270000      1.360000      1.072000
max    99998.000000     13.610000    772.330017     69.070000      5.460000

^Oh, almost forgot, we need to drop the first column of dataset.

## Eliminación de columnas no deseadas


In [ ]:
# La mejor manera de hacer esto en pandas es usar drop:
raw_data_na_dropped = raw_data_na_dropped.drop('Unnamed: 0', axis=1)

KeyError: ignored

In [ ]:
raw_data_na_dropped.describe()

Vmag           Plx         e_Plx           B-V
count  96742.000000  96742.000000  96742.000000  96742.000000
mean       8.291918      7.105242      1.264163      0.700663
std        1.247995     11.145643      1.156202      0.490053
min       -1.440000    -35.099998      0.380000     -0.400000
25%        7.610000      2.520000      0.870000      0.342000
50%        8.400000      4.600000      1.080000      0.605000
75%        9.070000      8.270000      1.360000      1.072000
max       13.610000    772.330017     69.070000      5.460000

In [ ]:
raw_data_na_dropped.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 96742 entries, 0 to 99998
Data columns (total 5 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Vmag    96742 non-null  float32
 1   Plx     96742 non-null  float32
 2   e_Plx   96742 non-null  float32
 3   B-V     96742 non-null  float32
 4   SpType  96742 non-null  object 
dtypes: float32(4), object(1)
memory usage: 3.0+ MB


Notamos que Int64Index tiene 96742 entradas, pero la indexación está de 0 a 99998.

Por lo tanto, necesitamos [reindexar](https://stackoverflow.com/questions/40755680/how-to-reset-index-pandas-dataframe-after-dropna-pandas-dataframe) nuestro marco de datos.



In [ ]:
raw_data_na_dropped_reindex = raw_data_na_dropped.reset_index(drop=True)

In [ ]:
raw_data_na_dropped_reindex.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 96742 entries, 0 to 96741
Data columns (total 5 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Vmag    96742 non-null  float32
 1   Plx     96742 non-null  float32
 2   e_Plx   96742 non-null  float32
 3   B-V     96742 non-null  float32
 4   SpType  96742 non-null  object 
dtypes: float32(4), object(1)
memory usage: 2.2+ MB


In [ ]:
#Optional - Save our progress
raw_data_na_dropped_reindex.to_csv("Star99999_na_dropped.csv", index=False)

# Creando una nueva columna para Amag

## Calculando la magnitu absoluta

The absolute magnitude of the stars were generated via the equation:

$$M = m+5(log_{10}p+1)$$

Donde $M$ representa la magnitud absoluta 'Amag',
**m** representa la magnitud visual aparente 'Vmag'
y **p** representa el paralaje estelar 'Plx'.


En esta sesión, crearemos una nueva columna ʻAmag` para almacenar $ M $.

Las cosas deben ser conscientes:
* Tomar logaritmo de 0 daría como resultado un infinito, que es lo que no queremos ver
  * Para solucionar esto: soltar filas con `Plx` = 0
* Tomar logaritmos de números negativos daría como resultado números complejos, que es lo que tampoco queremos ver.
  * Para arreglar esto: Tomando el valor absoluto de `Plx`


In [ ]:
#guardando una copia
df = raw_data_na_dropped_reindex.copy()
df

Vmag    Plx  e_Plx    B-V    SpType
0      9.10   3.54   1.39  0.482        F5
1      9.27  21.90   3.10  0.999       K3V
2      6.61   2.81   0.63 -0.019        B9
3      8.06   7.75   0.97  0.370       F0V
4      8.55   2.87   1.11  0.902     G8III
...     ...    ...    ...    ...       ...
96737  7.92   3.93   1.24  0.133     Am...
96738  8.72   3.07   0.87  0.097        B3
96739  8.08   1.07   0.68  1.094        G5
96740  6.98   2.97   0.76 -0.143     B1.5V
96741  8.51  -1.18   1.34  1.568  K5/M0III

[96742 rows x 5 columns]

In [ ]:
#Dropping rows that `Plx` = 0
df = df[df.Plx != 0] #ingresamos al data frame, los corchetes ingresan a una columna en específico

#Reindex the dataframe
df = df.reset_index(drop=True)

df

Vmag    Plx  e_Plx    B-V    SpType
0      9.10   3.54   1.39  0.482        F5
1      9.27  21.90   3.10  0.999       K3V
2      6.61   2.81   0.63 -0.019        B9
3      8.06   7.75   0.97  0.370       F0V
4      8.55   2.87   1.11  0.902     G8III
...     ...    ...    ...    ...       ...
96702  7.92   3.93   1.24  0.133     Am...
96703  8.72   3.07   0.87  0.097        B3
96704  8.08   1.07   0.68  1.094        G5
96705  6.98   2.97   0.76 -0.143     B1.5V
96706  8.51  -1.18   1.34  1.568  K5/M0III

[96707 rows x 5 columns]

Hemos removido todas las filas donde 'Plx' = 0.


In [ ]:
#Implementando la ecuación

df["Amag"] = df["Vmag"] + 5* (np.log10(abs(df["Plx"]))+1)

df

Vmag    Plx  e_Plx    B-V    SpType       Amag
0      9.10   3.54   1.39  0.482        F5  16.845016
1      9.27  21.90   3.10  0.999       K3V  20.972221
2      6.61   2.81   0.63 -0.019        B9  13.853533
3      8.06   7.75   0.97  0.370       F0V  17.506508
4      8.55   2.87   1.11  0.902     G8III  15.839410
...     ...    ...    ...    ...       ...        ...
96702  7.92   3.93   1.24  0.133     Am...  15.891963
96703  8.72   3.07   0.87  0.097        B3  16.155693
96704  8.08   1.07   0.68  1.094        G5  13.226919
96705  6.98   2.97   0.76 -0.143     B1.5V  14.343782
96706  8.51  -1.18   1.34  1.568  K5/M0III  13.869411

[96707 rows x 6 columns]

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 96707 entries, 0 to 96706
Data columns (total 6 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Vmag    96707 non-null  float32
 1   Plx     96707 non-null  float32
 2   e_Plx   96707 non-null  float32
 3   B-V     96707 non-null  float32
 4   SpType  96707 non-null  object 
 5   Amag    96707 non-null  float32
dtypes: float32(5), object(1)
memory usage: 2.6+ MB


In [ ]:
df.describe()

Vmag           Plx         e_Plx           B-V          Amag
count  96707.000000  96707.000000  96707.000000  96707.000000  96707.000000
mean       8.291762      7.107813      1.264152      0.700663     16.480989
std        1.247954     11.146737      1.156387      0.490001      2.451171
min       -1.440000    -35.099998      0.380000     -0.400000     -0.350000
25%        7.610000      2.520000      0.870000      0.342000     15.157670
50%        8.400000      4.600000      1.080000      0.605000     16.447716
75%        9.070000      8.270000      1.360000      1.072000     18.016191
max       13.610000    772.330017     69.070000      5.460000     30.449015

In [ ]:
#Save our dataset, we can finally play with it!!!
df.to_csv("stars_new_col.csv", index=False)

credits: https://www.kaggle.com/vinesmsuic/star-categorization-giants-and-dwarfs/notebooks

Puede usar este link para revisar más notebooks sobre este dataset:

https://www.kaggle.com/vinesmsuic/star-categorization-giants-and-dwarfs/notebooks
